File: C_Version/pid.c
#define NEUTRAL_ACTUATOR_POS 0.5 // steps/mm/etc where float is neutrally buoyant
#define ACTUATOR_MIN 0           // mechanical minimum
#define ACTUATOR_MAX 1           // mechanical maximum
#define BOUND 1000.0            // arbitrary integral windup bound
typedef struct
{
    double kp;
    double ki;
    double kd;
    double dt;
    double integral;
    double prev_error;
    double output_min;
    double output_max;
} PIDController;

void pid_init(PIDController *pid, double kp, double ki, double kd, double dt, double output_min, double output_max)
{
    pid->kp = kp;
    pid->ki = ki;
    pid->kd = kd;
    pid->dt = dt;
    pid->integral = 0.0;
    pid->prev_error = 0.0;
    pid->output_min = output_min;
    pid->output_max = output_max;
}

void pid_update(PIDController *pid, double error, double *output)
{
    // Proportional term
    double P = pid->kp * error;

    // Integral term
    pid->integral += error * pid->dt;

    // Simple anti-windup: clamp the internal integral sum
    if (pid->integral > BOUND)
    {
        pid->integral = BOUND;
    }
    if (pid->integral < -BOUND)
    {
        pid->integral = -BOUND;
    }
        
    double I = pid->ki * pid->integral;

    // Derivative term
    double derivative = (error - pid->prev_error) / pid->dt;
    double D = pid->kd * derivative;

    // Compute total output
    double raw_output = P + I + D;

    // Clamp output to min/max
    if (raw_output > pid->output_max)
    {
        raw_output = pid->output_max;
    }
    else if (raw_output < pid->output_min)
    {
        raw_output = pid->output_min;
    }

    // Update previous error
    pid->prev_error = error;

    // Set output
    *output = raw_output;
}

void pid_reset(PIDController *pid)
{
    pid->integral = 0.0;
    pid->prev_error = 0.0;
}

File: Python_Version/pid.py
import time

NEUTRAL_ACTUATOR_POS = 0.5  # steps/mm/etc where float is neutrally buoyant
ACTUATOR_MIN = 0            # mechanical minimum
ACTUATOR_MAX = 1            # mechanical maximum

class PID:
    def __init__(self, kp = 50.0, ki = 0.5, kd = 5.0, dt = 0.1, output_min = -0.5, output_max = 0.5):
        self.kp = kp
        self.ki = ki
        self.kd = kd
        self.dt = dt

        self.target_depth = 0
        
        self.integral = 0.0
        self.prev_error = 0.0

        self.output_min = output_min
        self.output_max = output_max

    def reset(self):
        self.integral = 0.0
        self.prev_error = 0.0

    def update(self, error):
        # Proportional term
        P = self.kp * error

        # Integral term (with basic anti-windup via clamping)
        self.integral += error * self.dt
        I = self.ki * self.integral

        # Derivative term
        derivative = (error - self.prev_error) / self.dt
        D = self.kd * derivative

        # PID output
        output = P + I + D

        # Store error for next derivative calc
        self.prev_error = error

        # Clamp output if limits given
        if self.output_min is not None:
            output = max(self.output_min, output)
        if self.output_max is not None:
            output = min(self.output_max, output)

        return output